File: elixir/2015/lib/day1.ex
defmodule Day1 do
  @spec parse(String.t() | charlist()) :: [-1 | 0 | 1]
  def parse([]), do: []
  def parse([head | tail]) do
    next =
      case head do
        ?( -> 1
        ?) -> -1
        _ -> 0
      end

    [next | parse(tail)]
  end
  def parse(s), do: s |> String.to_charlist() |> parse()

  @spec findPosition([{-1 | 0 | 1, integer()}]) :: integer()
  def findPosition([{v, i} | _]) when v == -1, do: i
  def findPosition([_ | tail]), do: findPosition(tail)

  @spec main() :: any()
  def main() do
    contents = File.read!("../../inputs/2015/Day1/input.txt")

    # part 1
    r1 = contents |> Day1.parse() |> Enum.sum()
    IO.inspect(r1)
    # part 2
    r2 =
      contents
      |> Day1.parse()
      |> Enum.scan(0, &(&1 + &2))
      |> Enum.with_index()
      |> Day1.findPosition()

    IO.inspect(r2 + 1)
  end
end

File: haskell/2015/Day1.hs
module Day1 where

import Data.List

parse :: Char -> Int
parse '(' = 1
parse ')' = -1
parse _ = 0

findPosition :: [Char] -> Int
findPosition xs = go $ zip [0 ..] $ scanl' (+) 0 $ map parse xs
  where
    go :: [(Int, Int)] -> Int
    go [] = error "did not find -1"
    go ((i, v) : xs) = if v == -1 then i else go xs

main' :: IO ()
main' = do
  contents <- head . lines <$> readFile "../inputs/2015/Day1/input.txt"
  -- part 1
  print $ sum $ map parse contents
  -- part 2
  print $ findPosition contents
